GADT with dependent type parameters causes `pickler crash  orphan parameter reference` when using Mirror

## Compiler version

3.7.2 

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```Scala
enum Expr {
  case UpcastToIterable[T, C <: Iterable[T]]()
}

object Expr {
  summon[scala.deriving.Mirror.SumOf[Expr]]
}
```

## Output (click arrow to expand)
<!--
Note: to get a full crash log when using scala-cli, run compilation with the flag `--server=false`
-->
<details>

```scala
 scalac expr_pickle_crash.scala
error when pickling type T
error when pickling type Iterable[T]
error when pickling type Expr.UpcastToIterable[Any, Iterable[T]]
error when pickling type Expr.UpcastToIterable[Any, Iterable[T]] *: EmptyTuple.type
error when pickling type  = Expr.UpcastToIterable[Any, Iterable[T]] *: EmptyTuple.type
error when pickling type scala.deriving.Mirror.Sum{
  type MirroredMonoType = Expr; type MirroredType = Expr;
    type MirroredLabel = ("Expr" : String);
    type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]] *:
      EmptyTuple.type
}
error when pickling type scala.deriving.Mirror.Sum{
  type MirroredMonoType = Expr; type MirroredType = Expr;
    type MirroredLabel = ("Expr" : String);
    type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]] *:
      EmptyTuple.type;
    type MirroredElemLabels = ("UpcastToIterable" : String) *: EmptyTuple.type
}
error when pickling tree 
  scala.deriving.Mirror.Sum{
    type MirroredMonoType = Expr; type MirroredType = Expr;
      type MirroredLabel = ("Expr" : String);
      type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]] *:
        EmptyTuple.type;
      type MirroredElemLabels = ("UpcastToIterable" : String) *: EmptyTuple.type
    }
 of class class dotty.tools.dotc.ast.Trees$InferredTypeTree
error when pickling tree val x$proxy1:
  
    scala.deriving.Mirror.Sum{
      type MirroredMonoType = Expr; type MirroredType = Expr;
        type MirroredLabel = ("Expr" : String);
        type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]] *:
          EmptyTuple.type;
        type MirroredElemLabels = ("UpcastToIterable" : String) *:
          EmptyTuple.type
    }
  
 =
  Expr.$asInstanceOf[
    
      scala.deriving.Mirror.Sum{
        type MirroredMonoType = Expr; type MirroredType = Expr;
          type MirroredLabel = ("Expr" : String);
          type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]] *:
            EmptyTuple.type;
          type MirroredElemLabels = ("UpcastToIterable" : String) *:
            EmptyTuple.type
      }
    
  ] of class class dotty.tools.dotc.ast.Trees$ValDef
error when pickling tree {
  val x$proxy1:
    
      scala.deriving.Mirror.Sum{
        type MirroredMonoType = Expr; type MirroredType = Expr;
          type MirroredLabel = ("Expr" : String);
          type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]] *:
            EmptyTuple.type;
          type MirroredElemLabels = ("UpcastToIterable" : String) *:
            EmptyTuple.type
      }
    
   =
    Expr.$asInstanceOf[
      
        scala.deriving.Mirror.Sum{
          type MirroredMonoType = Expr; type MirroredType = Expr;
            type MirroredLabel = ("Expr" : String);
            type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]] *:
              EmptyTuple.type;
            type MirroredElemLabels = ("UpcastToIterable" : String) *:
              EmptyTuple.type
        }
      
    ]
  x$proxy1
} of class class dotty.tools.dotc.ast.Trees$Inlined
error when pickling tree () extends Object(), scala.deriving.Mirror.Sum { this: Expr.type =>
  private def writeReplace(): AnyRef =
    new scala.runtime.ModuleSerializationProxy(classOf[Expr.type])
  final case class UpcastToIterable[T >: Nothing <: Any,
    C >: Nothing <: Iterable[T]]() extends Expr() {
    override def hashCode(): Int = -1772009433
    override def equals(x$0: Any): Boolean =
      (this eq x$0.$asInstanceOf[Object]) ||
        (x$0 match 
          {
            case x$0 @ _:Expr.UpcastToIterable[T, C] @unchecked => true
            case _ => false
          }
        )
    override def toString(): String = scala.runtime.ScalaRunTime._toString(this)
    override def canEqual(that: Any): Boolean =
      that.isInstanceOf[Expr.UpcastToIterable[T, C] @unchecked]
    override def productArity: Int = 0
    override def productPrefix: String = "UpcastToIterable"
    override def productElement(n: Int): Any =
      n match 
        {
          case _ => throw new IndexOutOfBoundsException(n.toString())
        }
    override def productElementName(n: Int): String =
      n match 
        {
          case _ => throw new IndexOutOfBoundsException(n.toString())
        }
    T
    C <: Iterable[T]
    def copy[T, C <: Iterable[T]](): Expr.UpcastToIterable[T, C] =
      new Expr.UpcastToIterable[T, C]()
    def ordinal: Int = 0
  }
  final lazy module val UpcastToIterable: Expr.UpcastToIterable =
    new Expr.UpcastToIterable()
  final module class UpcastToIterable() extends AnyRef(), 
    scala.deriving.Mirror.Product { this: Expr.UpcastToIterable.type =>
    private def writeReplace(): AnyRef =
      new scala.runtime.ModuleSerializationProxy(
        classOf[Expr.UpcastToIterable.type])
    def apply[T, C <: Iterable[T]](): Expr.UpcastToIterable[T, C] =
      new Expr.UpcastToIterable[T, C]()
    def unapply[T, C <: Iterable[T]](x$1: Expr.UpcastToIterable[T, C]): true =
      true
    override def toString: String = "UpcastToIterable"
    type MirroredMonoType = Expr.UpcastToIterable[? <: AnyKind, ? <: AnyKind]
    def fromProduct(x$0: Product): Expr.UpcastToIterable.MirroredMonoType =
      {
        new Expr.UpcastToIterable[Any, Iterable[Any]]()
      }
  }
  def fromOrdinal(ordinal: Int): Expr =
    throw 
      new NoSuchElementException(
        "enum Expr has no case with ordinal: " + ordinal.toString())
  {
    val x$proxy1:
      
        scala.deriving.Mirror.Sum{
          type MirroredMonoType = Expr; type MirroredType = Expr;
            type MirroredLabel = ("Expr" : String);
            type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]] *:
              EmptyTuple.type;
            type MirroredElemLabels = ("UpcastToIterable" : String) *:
              EmptyTuple.type
        }
      
     =
      Expr.$asInstanceOf[
        
          scala.deriving.Mirror.Sum{
            type MirroredMonoType = Expr; type MirroredType = Expr;
              type MirroredLabel = ("Expr" : String);
              type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]]
                *: EmptyTuple.type;
              type MirroredElemLabels = ("UpcastToIterable" : String) *:
                EmptyTuple.type
          }
        
      ]
    x$proxy1
  }
  type MirroredMonoType = Expr
  def ordinal(x$0: Expr.MirroredMonoType): Int = x$0.ordinal
} of class class dotty.tools.dotc.ast.Trees$Template
error when pickling tree @SourceFile("expr_pickle_crash.scala") final module class Expr() extends
  Object(), scala.deriving.Mirror.Sum { this: Expr.type =>
  private def writeReplace(): AnyRef =
    new scala.runtime.ModuleSerializationProxy(classOf[Expr.type])
  final case class UpcastToIterable[T >: Nothing <: Any,
    C >: Nothing <: Iterable[T]]() extends Expr() {
    override def hashCode(): Int = -1772009433
    override def equals(x$0: Any): Boolean =
      (this eq x$0.$asInstanceOf[Object]) ||
        (x$0 match 
          {
            case x$0 @ _:Expr.UpcastToIterable[T, C] @unchecked => true
            case _ => false
          }
        )
    override def toString(): String = scala.runtime.ScalaRunTime._toString(this)
    override def canEqual(that: Any): Boolean =
      that.isInstanceOf[Expr.UpcastToIterable[T, C] @unchecked]
    override def productArity: Int = 0
    override def productPrefix: String = "UpcastToIterable"
    override def productElement(n: Int): Any =
      n match 
        {
          case _ => throw new IndexOutOfBoundsException(n.toString())
        }
    override def productElementName(n: Int): String =
      n match 
        {
          case _ => throw new IndexOutOfBoundsException(n.toString())
        }
    T
    C <: Iterable[T]
    def copy[T, C <: Iterable[T]](): Expr.UpcastToIterable[T, C] =
      new Expr.UpcastToIterable[T, C]()
    def ordinal: Int = 0
  }
  final lazy module val UpcastToIterable: Expr.UpcastToIterable =
    new Expr.UpcastToIterable()
  final module class UpcastToIterable() extends AnyRef(), 
    scala.deriving.Mirror.Product { this: Expr.UpcastToIterable.type =>
    private def writeReplace(): AnyRef =
      new scala.runtime.ModuleSerializationProxy(
        classOf[Expr.UpcastToIterable.type])
    def apply[T, C <: Iterable[T]](): Expr.UpcastToIterable[T, C] =
      new Expr.UpcastToIterable[T, C]()
    def unapply[T, C <: Iterable[T]](x$1: Expr.UpcastToIterable[T, C]): true =
      true
    override def toString: String = "UpcastToIterable"
    type MirroredMonoType = Expr.UpcastToIterable[? <: AnyKind, ? <: AnyKind]
    def fromProduct(x$0: Product): Expr.UpcastToIterable.MirroredMonoType =
      {
        new Expr.UpcastToIterable[Any, Iterable[Any]]()
      }
  }
  def fromOrdinal(ordinal: Int): Expr =
    throw 
      new NoSuchElementException(
        "enum Expr has no case with ordinal: " + ordinal.toString())
  {
    val x$proxy1:
      
        scala.deriving.Mirror.Sum{
          type MirroredMonoType = Expr; type MirroredType = Expr;
            type MirroredLabel = ("Expr" : String);
            type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]] *:
              EmptyTuple.type;
            type MirroredElemLabels = ("UpcastToIterable" : String) *:
              EmptyTuple.type
        }
      
     =
      Expr.$asInstanceOf[
        
          scala.deriving.Mirror.Sum{
            type MirroredMonoType = Expr; type MirroredType = Expr;
              type MirroredLabel = ("Expr" : String);
              type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]]
                *: EmptyTuple.type;
              type MirroredElemLabels = ("UpcastToIterable" : String) *:
                EmptyTuple.type
          }
        
      ]
    x$proxy1
  }
  type MirroredMonoType = Expr
  def ordinal(x$0: Expr.MirroredMonoType): Int = x$0.ordinal
} of class class dotty.tools.dotc.ast.Trees$TypeDef
error when pickling tree package <empty> {
  @SourceFile("expr_pickle_crash.scala") @Child[Expr.UpcastToIterable] sealed
    abstract class Expr() extends Object(), scala.reflect.Enum {
    import Expr.UpcastToIterable
  }
  final lazy module val Expr: Expr = new Expr()
  @SourceFile("expr_pickle_crash.scala") final module class Expr() extends
    Object(), scala.deriving.Mirror.Sum { this: Expr.type =>
    private def writeReplace(): AnyRef =
      new scala.runtime.ModuleSerializationProxy(classOf[Expr.type])
    final case class UpcastToIterable[T >: Nothing <: Any,
      C >: Nothing <: Iterable[T]]() extends Expr() {
      override def hashCode(): Int = -1772009433
      override def equals(x$0: Any): Boolean =
        (this eq x$0.$asInstanceOf[Object]) ||
          (x$0 match 
            {
              case x$0 @ _:Expr.UpcastToIterable[T, C] @unchecked => true
              case _ => false
            }
          )
      override def toString(): String =
        scala.runtime.ScalaRunTime._toString(this)
      override def canEqual(that: Any): Boolean =
        that.isInstanceOf[Expr.UpcastToIterable[T, C] @unchecked]
      override def productArity: Int = 0
      override def productPrefix: String = "UpcastToIterable"
      override def productElement(n: Int): Any =
        n match 
          {
            case _ => throw new IndexOutOfBoundsException(n.toString())
          }
      override def productElementName(n: Int): String =
        n match 
          {
            case _ => throw new IndexOutOfBoundsException(n.toString())
          }
      T
      C <: Iterable[T]
      def copy[T, C <: Iterable[T]](): Expr.UpcastToIterable[T, C] =
        new Expr.UpcastToIterable[T, C]()
      def ordinal: Int = 0
    }
    final lazy module val UpcastToIterable: Expr.UpcastToIterable =
      new Expr.UpcastToIterable()
    final module class UpcastToIterable() extends AnyRef(), 
      scala.deriving.Mirror.Product { this: Expr.UpcastToIterable.type =>
      private def writeReplace(): AnyRef =
        new scala.runtime.ModuleSerializationProxy(
          classOf[Expr.UpcastToIterable.type])
      def apply[T, C <: Iterable[T]](): Expr.UpcastToIterable[T, C] =
        new Expr.UpcastToIterable[T, C]()
      def unapply[T, C <: Iterable[T]](x$1: Expr.UpcastToIterable[T, C]): true
         = true
      override def toString: String = "UpcastToIterable"
      type MirroredMonoType = Expr.UpcastToIterable[? <: AnyKind, ? <: AnyKind]
      def fromProduct(x$0: Product): Expr.UpcastToIterable.MirroredMonoType =
        {
          new Expr.UpcastToIterable[Any, Iterable[Any]]()
        }
    }
    def fromOrdinal(ordinal: Int): Expr =
      throw 
        new NoSuchElementException(
          "enum Expr has no case with ordinal: " + ordinal.toString())
    {
      val x$proxy1:
        
          scala.deriving.Mirror.Sum{
            type MirroredMonoType = Expr; type MirroredType = Expr;
              type MirroredLabel = ("Expr" : String);
              type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]]
                *: EmptyTuple.type;
              type MirroredElemLabels = ("UpcastToIterable" : String) *:
                EmptyTuple.type
          }
        
       =
        Expr.$asInstanceOf[
          
            scala.deriving.Mirror.Sum{
              type MirroredMonoType = Expr; type MirroredType = Expr;
                type MirroredLabel = ("Expr" : String);
                type MirroredElemTypes = Expr.UpcastToIterable[Any, Iterable[T]]
                   *: EmptyTuple.type;
                type MirroredElemLabels = ("UpcastToIterable" : String) *:
                  EmptyTuple.type
            }
          
        ]
      x$proxy1
    }
    type MirroredMonoType = Expr
    def ordinal(x$0: Expr.MirroredMonoType): Int = x$0.ordinal
  }
} of class class dotty.tools.dotc.ast.Trees$PackageDef

  unhandled exception while running pickler on expr_pickle_crash.scala

  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/scala/scala3/issues/new/choose
  For non-enriched exceptions, compile with -Xno-enrich-error-messages.


     while compiling: expr_pickle_crash.scala
        during phase: pickler
                mode: Mode(ImplicitsEnabled)
     library version: version 2.13.16
    compiler version: version 3.7.2
            settings: 

Exception in thread "main" java.lang.AssertionError: assertion failed: orphan parameter reference: TypeParamRef(T)
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:318)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:183)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$2(TreePickler.scala:204)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:204)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:183)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$2(TreePickler.scala:204)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:204)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:183)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$2(TreePickler.scala:204)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:204)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:183)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:279)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:183)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:271)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:183)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:270)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:183)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:716)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTpt(TreePickler.scala:352)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:382)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:647)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$11(TreePickler.scala:619)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:617)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:423)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:423)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:698)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:381)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:664)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:423)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:423)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:714)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle$$anonfun$1(TreePickler.scala:946)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:944)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$1$$anonfun$1(Pickler.scala:306)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:279)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.transform.Pickler.run(Pickler.scala:278)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:383)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:376)
	at dotty.tools.dotc.transform.Pickler.runPhase$1(Pickler.scala:392)
	at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:398)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:368)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1324)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:361)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$2(Run.scala:408)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$adapted$1(Run.scala:408)
	at scala.Function0.apply$mcV$sp(Function0.scala:42)
	at dotty.tools.dotc.Run.showProgress(Run.scala:470)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:408)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:420)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:420)
	at dotty.tools.dotc.Run.compileSources(Run.scala:307)
	at dotty.tools.dotc.Run.compile(Run.scala:292)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
	at dotty.tools.dotc.Driver.process(Driver.scala:201)
	at dotty.tools.dotc.Driver.process(Driver.scala:169)
	at dotty.tools.dotc.Driver.process(Driver.scala:181)
	at dotty.tools.dotc.Driver.main(Driver.scala:211)
	at dotty.tools.MainGenericCompiler$.run$1(MainGenericCompiler.scala:160)
	at dotty.tools.MainGenericCompiler$.main(MainGenericCompiler.scala:184)
	at dotty.tools.MainGenericCompiler.main(MainGenericCompiler.scala)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at coursier.bootstrap.launcher.a.a(Unknown Source)
	at coursier.bootstrap.launcher.Launcher.main(Unknown Source)
```
</details>

--- a/compiler/src/dotty/tools/dotc/typer/Synthesizer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Synthesizer.scala
@@ -598,8 +598,14 @@ class Synthesizer(typer: Typer)(using @constructorOnly c: Context):
                   resType <:< target
                   val tparams = poly.paramRefs
                   val variances = childClass.typeParams.map(_.paramVarianceSign)
-                  val instanceTypes = tparams.lazyZip(variances).map: (tparam, variance) =>
-                    TypeComparer.instanceType(tparam, fromBelow = variance < 0, Widen.Unions)
+                  @tailrec def fixInstances(cur: List[Type]): List[Type] =
+                    val next = cur.mapConserve(_.substParams(poly, cur))
+                    if next eq cur then next else fixInstances(next)
+                  val instanceTypes = {
+                    val types0 = tparams.lazyZip(variances).map: (tparam, variance) =>
+                      TypeComparer.instanceType(tparam, fromBelow = variance < 0, Widen.Unions)
+                    fixInstances(types0)
+                  }
                   val instanceType = resType.substParams(poly, instanceTypes)
                   // this is broken in tests/run/i13332intersection.scala,
                   // because type parameters are not correctly inferred.

--- a/tests/pos/i23774.scala
+++ b/tests/pos/i23774.scala
@@ -1,0 +1,35 @@
+trait Iterable[+A]
+
+enum Expr {
+  case Case[T, C <: Iterable[T]]()
+}
+object Expr {
+  val m = summon[scala.deriving.Mirror.SumOf[Expr]]
+  summon[m.MirroredElemTypes =:= Expr.Case[Any, Iterable[Any]] *: EmptyTuple]
+}
+
+enum Trait {
+  case Upcast[Base, Case <: Base]()
+}
+
+object Trait {
+  val m = summon[scala.deriving.Mirror.SumOf[Trait]]
+  summon[m.MirroredElemTypes =:= Trait.Upcast[Any, Any] *: EmptyTuple]
+}
+
+enum MoreDeps {
+  case Case[T, C <: Iterable[T], D <: Iterable[C]]()
+}
+object MoreDeps {
+  val m = summon[scala.deriving.Mirror.SumOf[MoreDeps]]
+  summon[m.MirroredElemTypes =:= MoreDeps.Case[Any, Iterable[Any], Iterable[Iterable[Any]]] *: EmptyTuple]
+}
+
+enum Circular {
+  case Case[C <: Iterable[D], D <: C]()
+}
+object Circular {
+  val m = summon[scala.deriving.Mirror.SumOf[Circular]]
+  summon[m.MirroredElemTypes =:= Circular.Case[Iterable[Nothing], Nothing] *: EmptyTuple]
+}
+

--- a/tests/pos/i23774typeclass.scala
+++ b/tests/pos/i23774typeclass.scala
@@ -1,0 +1,74 @@
+import scala.compiletime.{erasedValue, summonInline}
+import scala.deriving.Mirror
+
+enum Expr[+T]:
+  case UpcastToIterable[T, C <: Iterable[T]](v: Expr[C]) extends Expr[Iterable[T]]
+  case Seq[T](elements: Expr[T]*) extends Expr[scala.Seq[T]]
+  case Const(value: T)
+
+trait Fold[E]:
+  def apply[Acc](acc: Acc, expr: E, f: [t] => (Acc, Expr[t]) => Acc): Acc
+
+object Fold:
+  private inline def summonAll[Elems <: Tuple]: List[Fold[?]] =
+    inline erasedValue[Elems] match
+      case _: (h *: tail) => summonInline[Fold[h]] :: summonAll[tail]
+      case _: EmptyTuple => Nil
+
+  final class Leaf[E] extends Fold[E]:
+    def apply[Acc](acc: Acc, expr: E, f: [t] => (Acc, Expr[t]) => Acc): Acc = acc
+
+  given [T: Fold as fold] => Fold[Seq[T]] = new Fold[Seq[T]] {
+    def apply[Acc](acc: Acc, expr: Seq[T], f: [t] => (Acc, Expr[t]) => Acc): Acc =
+      expr.foldLeft(acc)((a, e) => fold(a, e, f))
+  }
+
+  given Fold[EmptyTuple] = new Fold[EmptyTuple]:
+    def apply[Acc](acc: Acc, expr: EmptyTuple, f: [t] => (Acc, Expr[t]) => Acc): Acc = acc
+
+  given [H: Fold as h, T <: Tuple: Fold as t] => Fold[H *: T] =
+    new Fold[H *: T]:
+      def apply[Acc](acc: Acc, expr: H *: T, f: [t] => (Acc, Expr[t]) => Acc): Acc =
+        val acc1 = h(acc, expr.head, f)
+        t(acc1, expr.tail, f)
+
+  private def product[E](m: Mirror.ProductOf[E], tupleFold: Fold[m.MirroredElemTypes]): Fold[E] =
+    new Fold[E]:
+      def apply[Acc](acc: Acc, expr: E, f: [t] => (Acc, Expr[t]) => Acc): Acc =
+        // The mirror makes this safe according to https://github.com/scala/scala3/issues/22382#issuecomment-2613187822
+        tupleFold(acc, Tuple.fromProduct(expr.asInstanceOf[Product]).asInstanceOf[m.MirroredElemTypes], f)
+
+  private def sum[E](m: Mirror.SumOf[E], cases0: () => List[Fold[?]]): Fold[E] =
+    new Fold[E]:
+      lazy val cases = cases0()
+      def apply[Acc](acc: Acc, expr: E, f: [t] => (Acc, Expr[t]) => Acc): Acc =
+        val ord = m.ordinal(expr)
+        val caseFold = cases.apply(ord)
+        caseFold.apply(acc, expr.asInstanceOf, f)
+
+  inline given derived[E](using m: Mirror.Of[E]): Fold[E] =
+    inline m match
+      case m: Mirror.SumOf[E] => sum(m, () => summonAll[m.MirroredElemTypes])
+      case m: Mirror.ProductOf[E] => product[E](m, summonInline[Fold[m.MirroredElemTypes]])
+
+  given [T] => Fold[Expr.Const[T]] = Leaf()
+  given Fold[Expr.UpcastToIterable[Any, Iterable[Any]]] = derived
+  given [T] => Fold[Expr[T]] = new Fold[Expr[T]]:
+    val default = derived[Expr[T]]
+    def apply[Acc](acc: Acc, expr: Expr[T], f: [t] => (Acc, Expr[t]) => Acc): Acc =
+      default(f(acc, expr), expr, f)
+
+@main def test(): Unit =
+  def count[T](expr: Expr[T], f: [t] => Expr[t] => Boolean)(using fold: Fold[Expr[T]]): Int =
+    fold(0, expr, [t] => (acc, e) => if f(e) then acc + 1 else acc)
+
+  val ast: Expr[Iterable[Int]] = Expr.UpcastToIterable(Expr.Seq(Expr.Const(1), Expr.Const(2), Expr.Const(3)))
+  val constCount = count(
+    ast,
+    [t] =>
+      _ match {
+        case Expr.Const(_) => true
+        case _ => false
+      }
+  )
+  println(s"Number of Const nodes: $constCount")

